File: examples/python/handoff.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import asyncio
import os
import ssl
import sys
from pathlib import Path

import certifi

# Add the project root to the Python path
sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from sandbox.sandbox import Sandbox

async def run_handoff(url_a: str, url_b: str):
    """
    Demonstrates the sandbox handoff functionality.

    To run this example:
    CLOUD_RUN_URL="wss://your-service-a-url.run.app" \
    CLOUD_RUN_URL_HANDOFF="wss://your-service-b-url.run.app" \
    python3 examples/python/handoff.py
    """
    print("--- Sandbox Handoff Example ---")
    print(f"Using URL A: {url_a}")
    print(f"Using URL B (Handoff): {url_b}")

    sandbox_a = None
    sandbox_b = None
    ssl_context = ssl.create_default_context(cafile=certifi.where())

    try:
        # 1. Create the first sandbox (Sandbox A) with handoff enabled
        print("\n--- Step 1: Creating Sandbox A ---")
        sandbox_a = await Sandbox.create(
            url=url_a,
            enable_sandbox_checkpoint=True,
            enable_sandbox_handoff=True,
            enable_debug=True,
            debug_label="SandboxA",
            ssl=ssl_context,
        )
        sandbox_id = sandbox_a.sandbox_id
        sandbox_token = sandbox_a.sandbox_token
        print(f"[SandboxA] Created with ID: {sandbox_id}")

        if not sandbox_id or not sandbox_token:
            raise RuntimeError("Sandbox A was created without an ID or token.")

        # Start a long-running process in Sandbox A
        print("[SandboxA] Starting a long-running process...")
        process_a = await sandbox_a.exec("python", 'import time; print("Process A running..."); time.sleep(5)')

        # Wait for the first chunk of output to confirm the process has started
        output_a = await process_a.stdout.read_all()
        print(f"[SandboxA] stdout: {output_a.strip()}")

        # 2. Attach to the same sandbox ID from a different client/URL (Sandbox B)
        print("\n--- Step 2: Attaching as Sandbox B to trigger handoff ---")
        sandbox_b = await Sandbox.attach(
            url=url_b,
            sandbox_id=sandbox_id,
            sandbox_token=sandbox_token,
            enable_debug=True,
            debug_label="SandboxB",
            ssl=ssl_context,
        )
        print(f"[SandboxB] Successfully attached to sandbox {sandbox_b.sandbox_id}. Handoff complete.")

        # 3. Verify Sandbox B is working by executing a command
        print("\n--- Step 3: Verifying Sandbox B is operational ---")
        process_b = await sandbox_b.exec("python", 'print("Hello from Sandbox B!")')
        
        output_b, _ = await asyncio.gather(
            process_b.stdout.read_all(),
            process_b.wait(),
        )
        print(f"[SandboxB] stdout: {output_b.strip()}")
        print("[SandboxB] Execution finished.")

        if "Hello from Sandbox B!" in output_b:
            print("\n✅ Handoff successful!")
        else:
            print("\n❌ Handoff failed: Did not receive expected output from Sandbox B.")

    except Exception as e:
        print(f"\n❌ An error occurred during the handoff process: {e}")
    finally:
        # Cleanup
        if sandbox_a:
            await sandbox_a.kill()
            print("[SandboxA] Killed.")
        if sandbox_b:
            await sandbox_b.kill()
            print("[SandboxB] Killed.")

def main():
    url_a = os.environ.get("CLOUD_RUN_URL")
    url_b = os.environ.get("CLOUD_RUN_URL_HANDOFF")

    if not url_a or not url_b:
        print("Error: Please set the CLOUD_RUN_URL and CLOUD_RUN_URL_HANDOFF environment variables.")
        print("Example: export CLOUD_RUN_URL=\"wss://your-service-a-url.run.app\"")
        print("         export CLOUD_RUN_URL_HANDOFF=\"wss://your-service-b-url.run.app\"")
        return

    # Convert http(s) URLs to ws(s) URLs
    url_a = url_a.replace("https://", "wss://").replace("http://", "ws://")
    url_b = url_b.replace("https://", "wss://").replace("http://", "ws://")

    asyncio.run(run_handoff(url_a, url_b))

if __name__ == "__main__":
    main()

File: examples/js/handoff.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * This example demonstrates the sandbox handoff functionality.
 *
 * It performs the following steps:
 * 1. Creates a new sandbox (Client A) on one server instance with handoff enabled.
 * 2. Starts a long-running process in Sandbox A.
 * 3. Attaches to the same sandbox ID from a different client (Client B), potentially
 *    on a different server instance.
 * 4. The server triggers a handoff: Sandbox A is checkpointed and its connection is closed.
 * 5. Client B successfully attaches and restores the sandbox from the checkpoint.
 * 6. Client B executes a command to verify it has taken over the session.
 *
 * SERVER-SIDE REQUIREMENTS:
 * To run this example, the Cloud Run service must be deployed with a persistent
 * volume (e.g., a GCS bucket) and the `CHECKPOINT_AND_RESTORE_PATH` environment
 * variable set to the mount path of that volume.
 *
 * This script expects the URLs of your deployed Cloud Run services to be
 * set in environment variables. For a local test, you can use the same URL for both.
 *
 * The WebSocket URLs should be in the format: wss://<your-cloud-run-url>
 *
 * To run this example:
 * 1. Make sure you have ts-node and uuid installed (`npm install -g ts-node && npm install uuid @types/uuid`).
 * 2. Set the environment variables:
 *    `export CLOUD_RUN_URL="wss://your-service-a-url.run.app"
 *    `export CLOUD_RUN_URL_HANDOFF="wss://your-service-b-url.run.app"
 * 3. Run the script from the root of the repository:
 *    `npx ts-node example/js/handoff.ts`
 */
import { Sandbox } from '../../clients/js/src/sandbox';
import { v4 as uuidv4 } from 'uuid';

async function main() {
  const urlA = process.env.CLOUD_RUN_URL;
  const urlB = process.env.CLOUD_RUN_URL_HANDOFF;

  if (!urlA || !urlB) {
    console.error("Error: Please set the CLOUD_RUN_URL and CLOUD_RUN_URL_HANDOFF environment variables.");
    console.error("Example: export CLOUD_RUN_URL=\"wss://your-service-a-url.run.app\"");
    console.error("         export CLOUD_RUN_URL_HANDOFF=\"wss://your-service-b-url.run.app\"");
    return;
  }

  console.log('--- Sandbox Handoff Example ---');
  console.log(`Using URL A: ${urlA}`);
  console.log(`Using URL B (Handoff): ${urlB}`);

  let sandboxA: Sandbox | null = null;
  let sandboxB: Sandbox | null = null;

  try {
    // 1. Create the first sandbox (Sandbox A) with handoff enabled
    console.log('\n--- Step 1: Creating Sandbox A ---');
    sandboxA = await Sandbox.create(urlA, {
      enableSandboxCheckpoint: true,
      enableSandboxHandoff: true,
      enableDebug: true,
      debugLabel: 'SandboxA',
    });
    const sandboxId = sandboxA.sandboxId;
    const sandboxToken = sandboxA.sandboxToken;
    console.log(`[SandboxA] Created with ID: ${sandboxId}`);

    if (!sandboxId || !sandboxToken) {
      throw new Error("Sandbox A was created without an ID or token.");
    }

    // Add a listener to see when Sandbox A's connection is closed
    sandboxA['connection'].on('close', (code: number) => {
      console.log(`[SandboxA] Connection closed with code: ${code}. This is expected during handoff.`);
    });

    // Start a long-running process in Sandbox A to ensure it's active
    console.log('[SandboxA] Starting a long-running process...');
    const processA = await sandboxA.exec('python', 'import time; print("Process A running..."); time.sleep(5)');
    
    // We can read the output iteratively. We'll wait for the first chunk to confirm
    // the process has started, then proceed to trigger the handoff.
    for await (const chunk of processA.stdout) {
      const output = chunk.toString().trim();
      console.log(`[SandboxA] stdout: ${output}`);
      if (output.includes("Process A running...")) {
        break;
      }
    }

    // 2. Attach to the same sandbox ID from a different client/URL (Sandbox B)
    // This will trigger the handoff process on the server.
    console.log('\n--- Step 2: Attaching as Sandbox B to trigger handoff ---');
    
    sandboxB = await Sandbox.attach(urlB, sandboxId, sandboxToken, {
        enableDebug: true,
        debugLabel: 'SandboxB',
    });
    console.log(`[SandboxB] Successfully attached to sandbox ${sandboxB.sandboxId}. Handoff complete.`);

    // 3. Verify Sandbox B is working by executing a command
    console.log('\n--- Step 3: Verifying Sandbox B is operational ---');
    const processB = await sandboxB.exec('python', 'print("Hello from Sandbox B!")');
    
    const [outputB] = await Promise.all([
      processB.stdout.readAll(),
      processB.wait(),
    ]);
    console.log(`[SandboxB] stdout: ${outputB.trim()}`);
    console.log('[SandboxB] Execution finished.');

    if (outputB.includes('Hello from Sandbox B!')) {
      console.log('\n✅ Handoff successful!');
    } else {
      console.error('\n❌ Handoff failed: Did not receive expected output from Sandbox B.');
    }

  } catch (error) {
    console.error('\n❌ An error occurred during the handoff process:', error);
  } finally {
    // Cleanup: ensure both sandbox connections are killed
    if (sandboxA) {
      await sandboxA.kill();
      console.log('[SandboxA] Killed.');
    }
    if (sandboxB) {
      await sandboxB.kill();
      console.log('[SandboxB] Killed.');
    }
  }
}

main().catch(console.error);
